doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,148 @@
+    )
+
+    # Write to custom MongoDB datasource.
+    # The args are passed to MongoDatasource.do_write().
+    ds.write_datasource(
+        MongoDatasource(), uri=MY_URI, database=MY_DATABASE, collection=MY_COLLECTION
+    )
+
+
+A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the 
+example here, let's call it ``MongoDatasource``. At a high level, it will have two 
+core parts to build out:
+
+* read support with :meth:`create_reader() <ray.data.Datasource.create_reader>`
+* write support with :meth:`do_write() <ray.data.Datasource.do_write>`.
+
+Here are the key design choices we will make in this guide:
+
+-  **MongoDB connector**: We use `PyMongo <https://pymongo.readthedocs.io/en/stable/>`__ to connect to MongoDB.
+-  **MongoDB to Arrow conversion**: We use `PyMongoArrow <https://mongo-arrow.readthedocs.io/en/latest/>`__ to convert query results into Arrow tables, which Datasets supports as a block format.
+-  **Parallel execution**: We ask the user to provide a list of MongoDB queries, with each corresponding to a read task (i.e. a :class:`~ray.data.ReadTask`) that Datasets will execute in parallel.
+
+------------
+Read support
+------------
+
+To support reading, we implement :meth:`create_reader() <ray.data.Datasource.create_reader>`, returning a :class:`~ray.data.Datasource.Reader` implementation for 
+MongoDB. This ``Reader`` creates a list of :class:`~ray.data.ReadTask` for the given 
+list of MongoDB queries. Each :class:`~ray.data.ReadTask` returns a list of blocks when called, and
+each :class:`~ray.data.ReadTask` is executed in remote workers to parallelize the execution.
+
+You can find documentation about :ref:`Ray Data blocks here <block-api>`.
+
+First, let's handle a single MongoDB query, which is the unit of execution in
+:class:`~ray.data.ReadTask`. We need to connect to MongoDB, execute the query against it,
+and then convert results into Arrow format. We use ``PyMongo`` and  ``PyMongoArrow``
+to achieve this.
+
+Read more about `the PyMongoArrow function here <https://mongo-arrow.readthedocs.io/en/stable/api/api.html#pymongoarrow.api.aggregate_arrow_all>`__.
+
+.. literalinclude:: ./doc_code/custom_datasource.py
+    :language: python
+    :start-after: __read_single_query_start__
+    :end-before: __read_single_query_end__
+
+Once we have this building block, we can just apply it for each of the provided MongoDB 
+queries. In particular, below, we construct a `_MongoDatasourceReader` by subclassing
+a :class:`~ray.data.Datasource.Reader`, and we implement `__init__` and `get_read_tasks`.
+
+In `__init__`, we pass in a couple arguments that will be eventually used in
+constructing the MongoDB query in `_read_single_query`.
+
+In `get_read_tasks`, we construct a `ReadTask` object for each `query` ("pipeline") object.
+A list of `ReadTask` objects are returned at the end of the function call, and these
+tasks are executed on remote workers by the Datasets execution engine (not shown).
```

Comment:
We should link to a relevant Key Concepts or feature guide section for this Datasets execution model call-out, e.g. this: https://docs.ray.io/en/master/data/key-concepts.html#reading-data.